lib/utils/templatize.js

```diff
@@ -402,19 +418,36 @@ function addPropagateEffects(template, templateInfo, options, methodHost) {
         warnOnUndeclaredProperties(templateInfo, options, methodHost);
       }
     }
-    upgradeTemplate(template, klass);
     // Mix any pre-bound data into __data; no need to flush this to
     // instances since they pull from the template at instance-time
-    if (template.__dataProto) {
+    if (target.__dataProto) {
       // Note, generally `__dataProto` could be chained, but it's guaranteed
       // to not be since this is a vanilla template we just added effects to
-      Object.assign(template.__data, template.__dataProto);
+      Object.assign(target.__data, target.__dataProto);
+    }
+    if (isTemplate) {
+      upgradeTemplate(target, klass);
+      // Clear any pending data for performance
+      target.__dataTemp = {};
+      target.__dataPending = null;
+      target.__dataOld = null;
+      target._enableProperties();
+    } else {
+      // Swizzle the cached subclass prototype onto the custom element
+      Object.setPrototypeOf(target, klass.prototype);
+      // Check for any pre-bound instance host properties, and do the
+      // instance property delete/assign dance for those (directly into data;
+      // not need to go through accessor since they are pulled at instance time)
+      const hostProps = templateInfo.hostProps;
+      for (let prop in hostProps) {
```

Comment:
When is this necessary?

Reply:
Bookmark: it seems like we could potentially not create accessors for these and instead mark them as having accessors so `setPendingProperty` is used and then we wouldn't have to remove the value here.

Reply:
I tested a modification to dom-if/dom-repeat that avoided creating forwarding accessors for host properties, and I couldn't measure a difference on `template-heavy`.

For posterity, here's the gambit I tested:
> I fudged `__dataAccessors` which is both what we use to avoid (re-)creating accessors, as well as to know whether to call `_setPendingProperty` or do assignments during binding.  However, this has to be done at a very critical time, after the dom-if/repeat receives its `_templateInfo` for its template (which contains the `hostProps` it needs to care about), but before bindings to it run.  The most sure thing was to make an accessor for `_templateInfo` on dom-if/repeat to trap when it is being set by the host to modify `__dataHasAccessor`:

```js
  set _templateInfo(info) {
    this.__templateInfo = info;
    const hostProps = info.hostProps;
    if (hostProps) {
      const accessors = this.__dataHasAccessor;
      for (const p in hostProps) {
        accessors['_host_' + p] = true;
      }
    }
  }
  get _templateInfo() {
    return this.__templateInfo;
  }
```

> After that, it was just a matter of ensuring a couple places (templatizer property initialization and event handling) read from `__data` rather than via the getter.  

The change avoided creating any accessors for host properties, and allowed me to delete the code here that does the instance property swizzle.  Again, not measurable.